File: V8-1/JT-Loader_Vis-Studio_C++/JTLoader/BitBuffer.cpp

#include "BitBuffer.h"

BitBuffer::BitBuffer(void)
{
	bitPos = 0;
	bitBuffer = 0x0000;
	nBits = 0;
	count = 0;
}

BitBuffer::~BitBuffer(void)
{
}

BitBuffer::BitBuffer(vector<UINT32> buffer)
{
	bitPos = 0;
	bitBuffer = 0x0000;
	nBits = 0;
	count = 0;
	m_buffer = buffer;
}

BitBuffer::BitBuffer(CodecDriver* pDriver)
{
	bitPos = 0;
	bitBuffer = 0x0000;
	nBits = 0;
	count = 0;
	_pDriver = pDriver;
}

int BitBuffer::getBitPos(void)
{
	int i = this->bitPos;
	return bitPos;
}

byte BitBuffer::readAsByte(int nbBits)
{
	return (byte)readAsLong(nbBits);
}

int BitBuffer::readAsInt(int nbBits)
{
	return (int)readAsLong(nbBits);
}

int BitBuffer::readAsInt(long bitPos, int nbBits)
{
	return (int)readAsLong(bitPos, nbBits);
}

long BitBuffer::readAsLong(int nbBits)
{
	return readAsLong(0, nbBits);
}

long BitBuffer::getBitBufBitSize(void)
{
    return m_buffer.size() * 8;
}

long BitBuffer::readAsLong(long bPos, int nbBits)
{
	long value = 0;
	long len = bPos + nbBits;

	// len = number of bits to read, we skip bPos bits and create a long
	// with nbBits bits
	while (len > 0)
	{
		// Not enough bits in the buffer => We read another byte
		if (nBits == 0)
		{
			if (m_buffer.size() == 0)
			{
				bitBuffer = _pDriver->getNextCodeText();

				nBits = 8;
				bitBuffer &= (int)0xFFL;

				count = (count + 1) % _pDriver->numContexts();
			}
			else
			{
				bitBuffer = m_buffer.at(count);

				nBits = 8;
				bitBuffer &= (int)0xFFL;

				count = (count + 1) % m_buffer.size();
			}
		}

		// This test skips the first bPos bits
		if (bPos == 0)
		{
			value <<= 1;
			// The value of the msb is added to the value result
			int test = bitBuffer >> 7;
			value |= (int)(bitBuffer >> 7);
		}
		else
		{
			bPos--;
		}
		// Remove the msb so the 2nd bit becomes the msb
		bitBuffer <<= 1;
		bitBuffer &= (int)0xFFL;
		nBits--;
		len--;
		bitPos++;

	}

	return value;
}

File: V8-1/JT-Loader_Vis-Studio_C#/JTLoader/Codecs/BitBuffer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace C_sharp_JT_Reader
{
    public class BitBuffer
    {
        int bitBuffer;  // Temporary i/o buffer
        int nBits;      // Number of bits in bitBuffer
        int bitPos;
        int count;

        Int32[] buffer;

        public BitBuffer(Int32[] buffer)
        {
            this.buffer = buffer;
            this.bitPos = 0;
            this.bitBuffer = 0x0000;
            this.nBits = 0;
            this.count = 0;
        }

        public int getBitPos()
        {
            return bitPos;
        }

        public byte readAsByte(int nbBits)
        {
            return (byte)readAsLong(nbBits);
        }

        public int readAsInt(int nbBits)
        {
            return (int)readAsLong(nbBits);
        }

        public int readAsInt(long bitPos, int nbBits)
        {
            return (int)readAsLong(bitPos, nbBits);
        }
        
        public long getBitBufBitSize()
        {
            //return this.buffer.limit() * 8;

            int test = this.buffer.Length * 8;

            return this.buffer.Length * 8;//////////////////////////////////////////////////////////////////////////////////// Check !! //////////////////////////////////////
        }

        public long readAsLong(int nbBits)
        {
            return readAsLong(0, nbBits);
        }

        // Read specified number of bits (max 32) starting from the given bit position, return the value as long.
        public long readAsLong(long bPos, int nbBits)
        {
            long value = 0;
            long len = bPos + nbBits;

            // len = number of bits to read, we skip bPos bits and create a long
            // with nbBits bits
            while (len > 0)
            {
                // Not enough bits in the buffer => We read another byte
                if (this.nBits == 0)
                {
                    bitBuffer = (int)buffer[count];
                    this.nBits = 8;
                    bitBuffer &= (int)0xFFL;

                    this.count = (this.count + 1) % buffer.Length;
                }

                // This test skips the first bPos bits
                if (bPos == 0)
                {
                    value <<= 1;
                    // The value of the msb is added to the value result
                    int test = bitBuffer >> 7;
                    value |= (int)(bitBuffer >> 7);
                }
                else
                {
                    bPos--;
                }
                // Remove the msb so the 2nd bit becomes the msb
                bitBuffer <<= 1;
                bitBuffer &= (int)0xFFL;
                this.nBits--;
                len--;
                this.bitPos++;

            }

            return value;
        }
    }
}
